cardano-dex-contracts-onchain/ErgoDex/PContracts/PPool.hs

```diff
@@ -193,6 +201,62 @@ findPoolOutput =
                 )
                 (const $ ptraceError "Pool output not found")
 
+poolCheckStakeChange :: 
+    ClosedTerm 
+        (    PBuiltinList (PAsData PPubKeyHash)
+        :--> PBuiltinList (PAsData PPubKeyHash)
+        :--> PInteger 
+        :--> PInteger 
+        :--> PInteger 
+        :--> PoolConfig 
+        :--> PDatum
+        :--> PAddress 
+        :--> PAddress 
+        :--> PBool
+        )
+poolCheckStakeChange = plam $ \stakeChangers signatories dx dy dlq prevDatum newDatumPData prevAddress newAddress -> unTermCont $ do
+    PDatum succD'' <- pmatchC newDatumPData 
+    succD' <- tlet $ ptryFromData @(PoolConfig) $ succD''    
+    prevCred <- tletField @"credential" prevAddress
+    newCred  <- tletField @"credential" newAddress
+    prevConf <- pletFieldsC @'["poolNft", "poolX", "poolY", "poolLq", "feeNum"] prevDatum
+    newConf  <- pletFieldsC @'["poolNft", "poolX", "poolY", "poolLq", "feeNum"] succD'
+    let
+        prevPoolNft    = getField @"poolNft" prevConf
+        prevPoolX      = getField @"poolX"   prevConf
+        prevPoolY      = getField @"poolY"   prevConf
+        prevPoolLq     = getField @"poolLq"  prevConf
+        prevPoolFeeNum = getField @"feeNum"  prevConf
+
+        newPoolNft    = pfromData $ getField @"poolNft" newConf
+        newPoolX      = pfromData $ getField @"poolX"   newConf
+        newPoolY      = pfromData $ getField @"poolY"   newConf
+        newPoolLq     = pfromData $ getField @"poolLq"  newConf
+        newPoolFeeNum = pfromData $ getField @"feeNum"  newConf
+
+        validPoolParams = 
+            prevPoolNft    #== newPoolNft    #&&
+            prevPoolX      #== newPoolX      #&&
+            prevPoolY      #== newPoolY      #&&
+            prevPoolLq     #== newPoolLq     #&&
+            prevPoolFeeNum #== newPoolFeeNum
+
+        validDelta = dx #== 0 #&& dy #== 0 #&& dlq #== 0
+        validCred  = prevCred #== newCred
+        
+    let 
+        stakeChanger   = pfromData $ phead # stakeChangers
+        validSignature = containsSignature # signatories # stakeChanger
+    
+    pure $ validDelta #&& validPoolParams #&& validCred #&& validSignature
+
+poolCheckAmmAction :: ClosedTerm (PoolConfig :--> PoolState :--> PInteger :--> PInteger :--> PInteger :--> PBool :--> PBool)
+poolCheckAmmAction = plam $ \conf s0 dx dy dlq isSwap ->
+    pif
+       isSwap
+       (dlq #== 0 #&& validSwap # conf # s0 # dx # dy)
+       (validDepositRedeem # s0 # dx # dy # dlq)
```

Comment:
Wtf, you just multiply terms by moving simple `if` branching into a function. Then you call this function twice in `Swap` and `_` branches.